packages/collector/src/index.ts

```diff
@@ -1 +1,211 @@
+  const latestBlock = await provider.getBlockNumber()
+  // current block step to fetch events
+  const blockStep = 45000 // adjust as per your RPC limitations
+
+  let allEvents: ethers.Event[] = []
+
+  // if the start block is not found in Redis, get events from the block where the market was deployed to the latest block
+  // or if the difference between the latest block and the start block is greater than the block step
+  if (!startBlock || latestBlock - parseInt(startBlock) > blockStep) {
+    startBlock = startBlock || markets_block[marketName]
+
+    log(
+      `Getting events from block: ${chalk.green(startBlock)} to block: ${chalk.green(latestBlock)} for market: ${chalk.bold.blue(marketName)}`
+    )
+
+    // create an array to hold promises for each block range
+    const promises = []
+    for (let block = parseInt(startBlock); block < latestBlock; block += blockStep + 1) {
+      const fromBlock = block
+      const toBlock = Math.min(block + blockStep, latestBlock)
+      promises.push(ovlMarketContract.queryFilter('*', fromBlock, toBlock))
+    }
+
+    // execute all promises in parallel
+    const eventsArrays = await Promise.all(promises)
+    allEvents = eventsArrays.flat() // Flatten the array of arrays into a single array of events
+  } else {
+    // if the start block is found in Redis, get events from the last processed block to the latest block
+    log(
+      `Getting events from block: ${chalk.green(startBlock)} to block: ${chalk.green(latestBlock)} for market: ${chalk.bold.blue(marketName)}`
+    )
+    allEvents = await ovlMarketContract.queryFilter('*', parseInt(startBlock), latestBlock)
+  }
+
+  return allEvents
+}
+
+async function fetchAndProcessEventsForAllMarkets() {
+  if (!process.env.RPC_URLS) {
+    log(chalk.bold.red('At least one RPC_URLS must be provided'))
+    return
+  }
+
+  if (taskRunning) {
+    log(chalk.bold.red('Task is already running. Skipping this run...'))
+    return
+  }
+
+  taskRunning = true
+  log(chalk.bold.blue('Cron job started at:', new Date().toLocaleString()))
+  log('Collector module is running...')
+
+  try {
+    // due to the rate limits of the RPC provider, at the first run, we will fetch events for all markets running market by market
+    // after the first run, we will fetch events for all markets in parallel
+    const firstRun = !(await redis.get('FirstRun'))
+
+    if (firstRun) {
+      for (const [marketName, address] of Object.entries(markets)) {
+        const events = await fetchEvents(marketName)
+        await processEvents(marketName, events)
+      }
+
+      await redis.set('FirstRun', 'false')
+    } else {
+      const promises: Promise<unknown>[] = []
+
+      Object.entries(markets).forEach(([marketName, address]) => {
+        promises.push(
+          fetchEvents(marketName).then(async (events) => {
+            await processEvents(marketName, events)
+          })
+        )
+      })
+
+      await Promise.all(promises)
```

Comment:
Implementing allSettled in this part of the code is a good idea because here is processing all the markets, so if one of them fails, the others will be processed anyway.